synapse/lib/modelrev.py

```diff
@@ -245,6 +248,371 @@ async def save():
+                        )),
+                    )
+
+                    if len(nodeedits) >= 1000:
+                        await save()
+
+                if nodeedits:
+                    await save()
+
+    async def updateTagProps(self, tagprops, layers):
+
+        for layr in layers:
+
+            nodeedits = []
+            meta = {'time': s_common.now(), 'user': self.core.auth.rootuser.iden}
+
+            async def save():
+                await layr.storNodeEdits(nodeedits, meta)
+                nodeedits.clear()
+
+            for form, tag, prop in layr.getTagProps():
+
+                if form is None or prop not in tagprops:
+                    continue
+
+                tptyp = self.core.model.tagprops[prop]
+                stortype = tptyp.type.stortype
+
+                async for buid, valu in layr.iterTagPropRows(tag, prop, form=form):
+
+                    try:
+                        newval = tptyp.type.norm(valu)[0]
+                    except s_exc.BadTypeValu as e:
+                        oldm = e.errinfo.get('mesg')
+                        logger.warning(f'Bad prop value {tag}:{prop}={valu!r} : {oldm}')
+                        continue
+
+                    if newval == valu:
+                        continue
+
+                    nodeedits.append(
+                        (buid, form, (
+                            (s_layer.EDIT_TAGPROP_SET, (tag, prop, newval, None, stortype), ()),
+                        )),
+                    )
+
+                    if len(nodeedits) >= 1000:
+                        await save()
+
+            if nodeedits:
+                await save()
+
+    async def revModel20220202(self, layers):
+
+        (forms, props, tagprops) = self.getElementsByStortype(s_layer.STOR_TYPE_HUGENUM)
+
+        cnt = 0
+        layrmap = {layr.iden: layr for layr in layers}
+        nodeedits = {layr.iden: {'adds': [], 'dels': [], 'n2edges': []} for layr in layers}
+        meta = {'time': s_common.now(), 'user': self.core.auth.rootuser.iden}
+
+        async def save(buid, newbuid):
+
+            for layriden, edits in nodeedits.items():
+                if edits['adds']:
+                    await layrmap[layriden].storNodeEdits([(newbuid, form, edits['adds'])], meta)
+                    edits['adds'].clear()
+
+                if edits['dels']:
+                    await layrmap[layriden].storNodeEdits([(buid, form, edits['dels'])], meta)
+                    edits['dels'].clear()
+
+                if edits['n2edges']:
+                    await layrmap[layriden].storNodeEdits(edits['n2edges'], meta)
+                    edits['n2edges'].clear()
+
+        for form in forms:
+
+            ftyp = self.core.model.type(form)
+            stortype = ftyp.stortype
+
+            async for buid, sodes in self.core._liftByProp(form, None, layers):
+
+                valu = None
+                for _, sode in sodes:
+                    valu = sode.get('valu')
```

Comment:
Is there logic to mop these up when we change a BUID for an existing node with only ( for example ) tag data in the layer?

Reply:
Yep, we go through all the sodes here to find the valu so we can recompute the buid, then go through the sode for each layer and update all the tags/props/etc there using the new buid.